File: backend/app/api/v1/admin/businesses.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import List, Optional
from pydantic import BaseModel
from app.database import get_db
from app.models.user import User
from app.models.business import Business
from app.models.enums import BusinessType, SubscriptionPlan
from app.schemas.business import BusinessResponse
from app.utils.audit import log_action
from .deps import require_admin_role, require_super_admin

router = APIRouter()

@router.get("/businesses", response_model=List[BusinessResponse])
def list_all_businesses(
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin_role),
    search: Optional[str] = Query(None, description="Buscar por nombre"),
    business_type: Optional[BusinessType] = Query(None),
    plan_type: Optional[SubscriptionPlan] = Query(None),
    is_active: Optional[bool] = Query(None),
):
    query = db.query(Business)

    if search:
        query = query.filter(Business.name.ilike(f"%{search.lower()}%"))
    if business_type is not None:
        query = query.filter(Business.business_type == business_type)
    if plan_type is not None:
        query = query.filter(Business.plan_type == plan_type)
    if is_active is not None:
        query = query.filter(Business.is_active == is_active)

    return query.order_by(Business.created_at.desc()).all()

@router.get("/businesses/{business_id}", response_model=BusinessResponse)
def get_business_admin(
    business_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin_role)
):
    """Obtiene detalles de un negocio."""
    business = db.query(Business).filter(Business.id == business_id).first()
    if not business:
        raise HTTPException(404, "Negocio no encontrado")
    return business

class ModulesUpdate(BaseModel):
    module_inventory: bool | None = None
    module_sales: bool | None = None
    module_clients: bool | None = None
    module_portfolio: bool | None = None
    module_finance: bool | None = None
    module_suppliers: bool | None = None
    module_reports: bool | None = None
    module_waste: bool | None = None

@router.patch("/businesses/{business_id}/modules", response_model=BusinessResponse)
def update_business_modules(
    business_id: int,
    data: ModulesUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_super_admin)
):
    """
    Actualiza los módulos habilitados de un negocio (solo super_admin).
    Permite overrides manuales para dar acceso custom sin cambiar plan.
    """
    business = db.query(Business).filter(Business.id == business_id).first()
    if not business:
        raise HTTPException(404, "Negocio no encontrado")
    
    changes = {}
    for field, value in data.model_dump(exclude_none=True).items():
        old_value = getattr(business, field)
        if old_value != value:
            changes[field] = {"from": old_value, "to": value}
            setattr(business, field, value)
    
    db.commit()
    db.refresh(business)
    
    if changes:
        log_action(
            db, current_user.id, "UPDATE_BUSINESS_MODULES", "Business", business.id,
            details={"business_name": business.name, "changes": changes}
        )
    
    return business

File: frontend/src/api/admin/businesses.ts
import { api } from '../axios'
import type { Business, BusinessType, PlanType, SubscriptionPlan } from '@/types'

export interface AdminBusinessesFilters {
    search?: string
    business_type?: BusinessType
    plan_type?: SubscriptionPlan
    is_active?: boolean
}

export interface ModulesUpdate {
    module_inventory?: boolean
    module_sales?: boolean
    module_clients?: boolean
    module_portfolio?: boolean
    module_finance?: boolean
    module_suppliers?: boolean
    module_reports?: boolean
    module_waste?: boolean
}

export interface BusinessUpdateFull {
    name?: string
    description?: string
    business_type?: BusinessType
    plan_type?: PlanType
    max_users?: number
    max_products?: number
}

export interface BusinessUser {
    id: number
    user_id: number
    user_email: string
    user_name: string
    role_name: string | null
    is_active: boolean
    joined_at: string
}

export interface BusinessWithUsers extends Business {
    users?: BusinessUser[]
    owner_email?: string
    owner_name?: string
}

export const adminBusinessesApi = {
    list: (filters?: AdminBusinessesFilters) => {
        const params = new URLSearchParams()
        if (filters?.search) params.set('search', filters.search)
        if (filters?.business_type) params.set('business_type', filters.business_type)
        if (filters?.plan_type) params.set('plan_type', filters.plan_type)
        if (filters?.is_active !== undefined) params.set('is_active', String(filters.is_active))
        return api.get<Business[]>(`/admin/businesses?${params}`).then((r) => r.data)
    },

    getById: (businessId: number) =>
        api.get<BusinessWithUsers>(`/admin/businesses/${businessId}`).then((r) => r.data),

    updateModules: (businessId: number, data: ModulesUpdate) =>
        api.patch<Business>(`/admin/businesses/${businessId}/modules`, data).then((r) => r.data),

    updateFull: (businessId: number, data: BusinessUpdateFull) =>
        api.patch<Business>(`/admin/businesses/${businessId}/full`, data).then((r) => r.data),

    delete: (businessId: number) =>
        api.delete(`/admin/businesses/${businessId}`),
}